src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20230808_203400.java

```diff
@@ -0,0 +1,460 @@
+        long lastBatchExerciseCount = batchCountPerThread * BATCH_SIZE + exerciseCount % BATCH_SIZE;
+
+        for (int tc = 0; tc < threadCount; tc++) {
+            boolean isLastThread = tc == threadCount - 1;
+            long start = tc * batchCountPerThread;
+            long end = (tc + 1) * batchCountPerThread;
+            long exerciseCountForThread = batchCountPerThread * BATCH_SIZE;
+            executorService.submit(() -> {
+                Map<ProgrammingExercise, Boolean> threadErrorMap = new HashMap<>();
+                if (isLastThread) {
+                    executeBatchedMigration(start, lastBatchEnd, lastBatchExerciseCount, threadErrorMap);
+                }
+                else {
+                    executeBatchedMigration(start, end, exerciseCountForThread, threadErrorMap);
+                }
+                synchronized (errorMap) {
+                    errorMap.putAll(threadErrorMap);
+                }
+            });
+        }
+
+        // Wait for all threads to finish
+        executorService.shutdown();
+
+        try {
+            boolean finished = executorService.awaitTermination(3, TimeUnit.HOURS);
+            if (!finished) {
+                log.error(ERROR_MESSAGE);
+                throw new RuntimeException(ERROR_MESSAGE);
+            }
+        }
+        catch (InterruptedException e) {
+            log.error(ERROR_MESSAGE);
+            throw new RuntimeException(e);
+        }
+
+        evaluateErrorMap(errorMap);
+    }
+
+    /**
+     * Executes the migration for a batch of programming exercises. Expected to be in a separate thread.
+     *
+     * @param start         The start index of the batch page
+     * @param end           The end index of the batch page
+     * @param exerciseCount The total number of exercises for this thread
+     * @param errorMap      The error map to write errors to
+     */
+    private void executeBatchedMigration(long start, long end, long exerciseCount, Map<ProgrammingExercise, Boolean> errorMap) {
+        // do batches
+        for (int batchStartIndex = (int) start, batch = 0; batchStartIndex < end; batchStartIndex++, batch = batch + BATCH_SIZE) {
+            var batchStart = System.currentTimeMillis();
+            Pageable pageable = PageRequest.of(batchStartIndex, BATCH_SIZE);
+
+            // TODO: in the unlikely case the buildPlanId is null, we could skip the template participation
+            var templateParticipationPage = templateProgrammingExerciseParticipationRepository.findAll(pageable);
+            log.info("Found {} template programming exercises to migrate in batch", templateParticipationPage.getTotalElements());
+            templateParticipationPage.map(templateParticipation -> {
+                if (!ciMigrationService.orElseThrow().buildPlanExists(templateParticipation.getProgrammingExercise().getProjectKey(),
+                        templateParticipation.getProgrammingExercise().getTemplateBuildPlanId())) {
```

Comment:
Can you move the parameters to variables to that it can more easily be seen that this is a simple `buildPlanExists` check?